doc/terminology.rst

```diff
@@ -0,0 +1,51 @@
+.. _terminology:
+
+.. https://github.com/pydata/xarray/issues/2410
+.. https://github.com/pydata/xarray/issues/1295
+
+Terminology
+===========
+
+*Xarray terminology differs slightly from CF and mathematical conventions, and therefore using xarray, understanding the documentation, and parsing error messages is easier once key terminology is defined. This glossary was designed so that more fundamental concepts come first. Thus for new users, this page is best read top-to-bottom. Throughout the glossary,* ``arr`` *will refer to an xarray* :py:class:`DataArray` *in any small examples. For more complete examples, please consult the relevant documentation.*
+
+----
+
+**DataArray:** A multi-dimensional array with labeled or named dimensions. If its optional ``name`` property is set, it is a *named DataArray*.
+
+----
+
+**Dimension / dimensions:** A *dimension* is a nonnegative number for the dimensionality of the underlying data, while an array's *dimensions* are a set of dimension names. The name of the ``i``-th dimension is ``arr.dims[i]``. If an array is created without dimensions, the default dimension names are ``dim_0``, ``dim_1``, and so forth.
```

Comment:
I'm not sure about `A *dimension* is a nonnegative number for the dimensionality `—I think that's a good definition for dimensionality, but wouldn't think of a dimension as a number?

Reply:
I struggled with this because in math, a "dimension" is a number, but in common parlance and xarray, "dimension" is often used to refer to a _dimension axis_—or a locus in which all but one of the coordinates ("coordinates" in the math sense, not xarray sense) is fixed. I didn't want to define "coordinate" and "dimension" in the math sense and then re-define them, although I think that would be the clearest at the expense of brevity.

Anyway, how about this:

> **Dimension:** In mathematics, the *dimension* of data is loosely the number of degrees of freedom for it. A *dimension axis* is a set of all points in which all but one of these degrees of freedom is fixed. We can think of each dimension axis as having a name, for example the "x dimension".  In xarray, a `DataArray`'s *dimensions* are its named dimension axes, and the name of the ``i``-th dimension is ``arr.dims[i]``. If an array is created without dimensions, the default dimension names are ``dim_0``, ``dim_1``, and so forth.

Reply:
It might be worth borrowing the description from the netCDF data model documentation: https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_model.html

> Dimensions: describe the axes of the [DataArrays]. A dimension has a name and a length...